Documentation fixes (#202)
AY2122S1-CS2103T-W11-4/tp: docs/DeveloperGuide.md

**Target user profile**:

* property agents who need to manage a significant number of properties and buyers
* prefer desktop apps
* can type fast
* prefers typing to mouse interactions

> * is reasonably comfortable using CLI apps

Reviewer: Can we capitalise the first letter of the sentence?

* 1d. User requests to add a tag that is already present or delete a tag that is absent.
    * 1d1. `PropertyWhiz` warns the user about the tag's presence (in the case of add tag) or absence (in the case of delete tag).

    Use case resumes from step 2.

_________________

**Use case: UC05 - Find properties/buyers**

**MSS**

1. `User` requests to find properties/buyers by specifying keywords, tags or price range.
2. `PropertyWhiz` displays properties/buyers that fulfill the criteria.

Use case ends.

_________________

**Use case: UC06 - Import properties/buyers**

**MSS**

1. `User` requests to import properties/buyers.
2. `User` chooses the source CSV file.
3. `PropertyWhiz` imports the properties/buyers from the source file.

Use case ends.

**Extensions**

* 1a. `User` cancels while choosing a CSV file.
    * 1a1. `PropertyWhiz` shows an error message.

    Use case ends.

* 2a. The CSV file cannot be opened or read from.
    * 2a1. `PropertyWhiz` shows an error message.

    Use case ends.

* 2b. The CSV file is incorrectly formatted.
    * 2b1. `PropertyWhiz` shows an error message.

    Use case ends.

> * 2c. Importing creates duplicates within `PropertyWhiz`.

Reviewer: ```suggestion
* 2c. Importing causes duplicates within `PropertyWhiz`.
```

Author: I think creates sounds better to me, what do the others think? @AY2122S1-CS2103T-W11-4/developers

    * 2c1. `PropertyWhiz` shows an error message.

    Use case ends.

_________________

**Use case: UC07 - Export properties/buyers**

**MSS**

1. `User` requests to export properties/buyers.
2. `User` chooses the destination CSV file.
3. `PropertyWhiz` exports displayed properties/buyers to the CSV file.

Use case ends.

**Extensions**

* 1a. `User` cancels while choosing a CSV file.

>     * 1a1. `PropertyWhiz` shows an error message.

Reviewer 2: The message "Import properties was cancelled." isn't really an error messge, maybe "`PropertyWhiz` reports that import is cancelled"

    Use case ends.

* 2a. The CSV file cannot be opened or written to.
    * 2a1. `PropertyWhiz` shows an error message.

    Use case ends.

_________________

**Use case: UC08 - Match auto**

**MSS**

1. `User` requests to match properties to buyers 1-to-1.
2. `PropertyWhiz` displays 1-to-1 matching between properties and buyers.

Use case ends.

**Extensions**

* 1a. There are either no properties, or no buyers.
    * 1a1. `PropertyWhiz` shows an error message.

    Use case ends.

_________________

**Use case: UC09 - Charting**

**MSS**

> 1. `User` requests to for a chart of prices of properties/buyers.
> 2. `PropertyWhiz` draws and displays chart showing prices of properties/buyers.

Reviewer: ```suggestion
1. `User` requests for a chart of prices of properties/buyers.
2. `PropertyWhiz` draws and displays chart showing prices of properties/buyers in a new window.
```

Author: Ui details though

Use case ends.

### Non-Functional Requirements

1. Should work on any _mainstream OS_ as long as it has Java `11` or above installed.
2. Should be able to hold up to 200 properties and buyers without a noticeable sluggishness in performance for typical usage.
3. A user with above average typing speed for regular English text (i.e. not code, not system admin commands) should be able to accomplish most of the tasks faster using commands than using the mouse.
4. Should be able to function without internet access.
5. Should be able to restart without loss of data.

> 6. Should be packaged within a single jar, without requiring an installer.

Reviewer 2: "single JAR file"

1. Initial launch

   1. Download the jar file and copy into an empty folder

>    2. Double-click the jar file Expected: Shows the GUI with a set of sample properties and buyers. The window size may not be optimum.

Reviewer 2: should uppercase all "jar" i.e. JAR to be consistent with "csv"

   1. Prerequisites: List all properties using the `list` command. Multiple properties in the list.

   2. Test case: `delete property 1`<br>
      Expected: First property is deleted from the list. Details of the deleted property shown.

   3. Test case: `delete property 0`<br>
      Expected: No property is deleted. Error message shown.

   4. Other incorrect delete commands to try: `delete`, `delete property x` (where x is larger than the list size)<br>
      Expected: Similar to previous test cases.

4. Sort properties that are in the displayed list
   1. Test case: `sort property name desc`<br>

>       Expected: Sorted all properties in the displayed properties list by name in descending order.

Reviewer: ```suggestion
      Expected: Sorted all properties in the displayed property list by name in descending order.
```

   2. Test case: `sort property name`<br>

>       Expected: As there is no sort direction given, properties are not sorted.

Reviewer 3: This should be "Error message shown"?

### Exporting and importing data

1. Testing properties export/import CSV
   1. Remove `propertywhiz.json` file in the /data folder.
   2. Run `PropertyWhiz` to obtain an initialized list of properties and buyers.

>    3. Enter Command: `export properties`, select a CSV file at a convenient location.<br>

Reviewer 2: `export properties` -> `export property` to be consistent with UG

      Expected: CSV file is created, containing a header and 1 property per row.
   4. Enter Command: `clear` to clear existing properties and buyers.
   5. Enter Command: `import properties` from the CSV file.<br>
      Expected: Properties are successfully imported.
   6. Enter Command: `import properties` from the CSV file.<br>
      Expected: Import properties failed due to duplicates.

> 2. Repeat the commands above to test for buyers.

Reviewer 2: For the section **Testing multiple commands quickly**, add one more `back` command at line 808

  * Command parers have to work for both `Property` and `Buyer`. We considered several user command formats that can work with both `Property` and `Buyer`,
  and settled on adding `property/properties`, `buyer/buyers` after a command word. 
  * Separate command logic and test, storage is required for `Property` and `Buyer` classes. This greatly increased the workload of implementation and testing. 
* Import and export CSV file commands
  * A new workflow was required for the UI to handle files. We must first parse the command text to check if it requires a file, request for a file if needed, then consider executing the command. 
  * Solution: we added a new type of command- `CommandWithFile` to represent commands that require a file. We then added `CommandPreAction` to indicate that a FileChooserDialog is required for the user to select a file. 

>   * `CsvManager` was added as a static, purely functional class. The parsing of CSV file was tricky and involved in multiple exceptions. Careful analysis of cases was required to display an appropriate error message.

Reviewer 2: ... CSV file**s** was tricky and was involved in multiple...
